plugin/rename_file.py

```diff
@@ -0,0 +1,142 @@
+
+    def name(self) -> str:
+        return "new_name"
+
+    def placeholder(self) -> str:
+        return self.path
+
+    def initial_text(self) -> str:
+        return self.placeholder()
+
+    def initial_selection(self) -> list[tuple[int, int]]:
+        name, _ext = os.path.splitext(self.path)
+        return [(0, len(name))]
+
+    def validate(self, path: str) -> bool:
+        return len(path) > 0
+
+
+class LspRenamePathCommand(LspWindowCommand):
+    capability = 'workspace.fileOperations.willRename'
+
+    def is_enabled(self) -> bool:
+        return True
+
+    def want_event(self) -> bool:
+        return False
+
+    def input(self, args: dict) -> sublime_plugin.TextInputHandler | None:
+        if "new_name" in args:
+            return None
+        view = self.window.active_view()
+        old_path = args.get('old_path') or view.file_name() if view else None
+        return RenamePathInputHandler(Path(old_path or "").name)
+
+    def run(self, new_name: str, old_path: str | None = None) -> None:
+        view = self.window.active_view()
+        old_path = old_path or view.file_name() if view else None
+        if old_path is None:  # handle renaming buffers
+            if view:
+                view.set_name(new_name)
+            return
+        # new_name can be: FILE_NAME.xy OR ./FILE_NAME.xy OR ../../FILE_NAME.xy
+        resolved_new_path = (Path(old_path).parent / new_name).resolve()
+        if resolved_new_path.exists():
+            self.window.status_message('Unable to Rename. Already exists')
+            return
+        new_path = str(resolved_new_path)
+        new_path_uri = filename_to_uri(new_path)
+        old_path_uri = filename_to_uri(old_path)
+        rename_file_params: RenameFilesParams = {
+            "files": [{
+                "newUri": new_path_uri,
+                "oldUri": old_path_uri
+            }]
+        }
+        session = self.session()
+        file_operation_options = session.get_capability('workspace.fileOperations.willRename') if session else None
+        if session and file_operation_options and match_file_operation_filters(file_operation_options, old_path_uri):
+            request = Request.willRenameFiles(rename_file_params)
+            session.send_request(
+                request,
+                lambda response: self._handle_response_async(response, session.config.name,
+                                                             old_path, new_path, rename_file_params)
+            )
+        else:
+            self.rename_path(old_path, new_path)
+            self.notify_did_rename(rename_file_params)
+
+    def _handle_response_async(self, response: WorkspaceEdit | None, session_name: str,
+                               old_path: str, new_path: str, rename_file_params: RenameFilesParams) -> None:
+        def on_workspace_edits_applied(_) -> None:
+            self.rename_path(old_path, new_path)
+            self.notify_did_rename(rename_file_params)
```

Comment:
`notify_did_rename` should only be called if rename doesn't fail.
And could wait for promises in `rename_path` also (even if not critical).

Reply:
[9f667a3](https://github.com/sublimelsp/LSP/pull/2498/commits/9f667a38b59b1c523e8ef0c8ac0fe2d4c2aeea22)